plan/expression_rewriter.go

```diff
@@ -245,26 +246,196 @@ func (er *expressionRewriter) handleCompareSubquery(v *ast.CompareSubqueryExpr)
+		FromID:   agg.id,
+		Position: posID,
+		RetType:  countNull.GetType(),
+	}
+	aggColCount := &expression.Column{
+		ColName:  model.NewCIStr("agg_col_cnt"),
+		FromID:   agg.id,
+		Position: posID + 1,
+		RetType:  count.GetType(),
+	}
+	agg.schema.Append(aggColCountNull)
+	agg.schema.Append(aggColCount)
+	if all {
+		// All of the inner record set should not contain null value. So for t.id < all(select s.id from s), it
+		// should be rewrited to t.id < min(s.id) and if(count(s.id) == count(s.id is null), true, null).
+		countEQ, _ := expression.NewFunction(ast.EQ, types.NewFieldType(mysql.TypeTiny), aggColCountNull.Clone(), aggColCount.Clone())
+		nullChecker, _ := expression.NewFunction(ast.If, types.NewFieldType(mysql.TypeTiny), countEQ, expression.One, expression.Null)
+		cond = expression.ComposeCNFCondition(cond, nullChecker)
+		// If the set is empty, it should always return true.
+		checkEmpty, _ := expression.NewFunction(ast.EQ, types.NewFieldType(mysql.TypeTiny), aggColCountNull.Clone(), &expression.Constant{
+			Value:   types.NewDatum(0),
+			RetType: types.NewFieldType(mysql.TypeShort),
+		})
+		cond = expression.ComposeDNFCondition(cond, checkEmpty)
+	} else {
+		// For "any" expression, if the record set has null and the cond return false, the result should be NULL.
+		countNE, _ := expression.NewFunction(ast.NE, types.NewFieldType(mysql.TypeTiny), aggColCountNull.Clone(), aggColCount.Clone())
+		nullChecker, _ := expression.NewFunction(ast.If, types.NewFieldType(mysql.TypeTiny), countNE, expression.Null, expression.Zero)
+		cond = expression.ComposeDNFCondition(cond, nullChecker)
+	}
+	if !er.asScalar {
+		// For Semi Apply without aux column, the result is no matter false or null. So we can add it to join predicate.
+		er.p = er.b.buildSemiApply(er.p, agg, []expression.Expression{cond}, false, false)
+		return
+	}
+	// If we treat the result as a scalar value, we will add a projection with a extra column to output true, false or null.
+	outerSchemaLen := er.p.GetSchema().Len()
+	er.p = er.b.buildInnerApply(er.p, agg)
+	joinSchema := er.p.GetSchema()
+	proj := &Projection{
+		baseLogicalPlan: newBaseLogicalPlan(Proj, er.b.allocator),
+		Exprs:           expression.Column2Exprs(joinSchema.Clone().Columns[:outerSchemaLen]),
+	}
+	proj.self = proj
+	proj.initIDAndContext(er.ctx)
+	proj.SetSchema(expression.NewSchema(joinSchema.Clone().Columns[:outerSchemaLen]))
+	proj.Exprs = append(proj.Exprs, cond)
+	proj.schema.Append(&expression.Column{
+		FromID:      proj.id,
+		ColName:     model.NewCIStr("aux_col"),
+		Position:    proj.schema.Len(),
+		IsAggOrSubq: true,
+		RetType:     cond.GetType(),
+	})
+	addChild(proj, er.p)
+	er.p = proj
+}
+
+// handleNEAny handles the case of != any. For exmaple, if the query is t.id != any (select s.id from s), it will be rewrited to
+// t.id != s.id or count(distinct s.id) > 1. If there are two different values in s.id , there must exist a s.id that doesn't equal to t.id.
```

Comment:
sum(t.id = s.id) will make t.id as correlated column, but aggregation can be decorrelated now.